system/lib/wasmfs/wasmfs.cpp

```diff
@@ -153,90 +234,138 @@ long __syscall_fstat64(long fd, long buf) {
-                              const __wasi_iovec_t* iovs,
-                              size_t iovs_len,
-                              __wasi_size_t* nread) {
-  auto openFile = FileTable::get()[fd];
+__wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
+  int accessMode = (flags & O_ACCMODE);
+  bool canWrite = false;
 
-  if (!openFile) {
-    return __WASI_ERRNO_BADF;
+  if (accessMode == O_WRONLY || accessMode == O_RDWR) {
+    canWrite = true;
   }
 
-  auto file = openFile.locked().getFile()->dynCast<DataFile>();
+  // TODO: remove assert when all functionality is complete.
+  assert(((flags) & ~(O_CREAT | O_EXCL | O_DIRECTORY | O_TRUNC | O_APPEND |
+                      O_RDWR | O_WRONLY | O_RDONLY | O_LARGEFILE)) == 0);
 
-  // If file is nullptr, then the file was not a DataFile.
-  if (!file) {
-    return __WASI_ERRNO_ISDIR;
+  std::vector<std::string> pathParts = splitPath((char*)pathname);
+
+  if (pathParts.empty()) {
+    return -(EINVAL);
   }
 
-  auto lockedFile = file->locked();
+  // Root directory
+  if (pathParts.size() == 1 && pathParts[0] == "/") {
+    auto openFile =
+      std::make_shared<OpenFileState>(0, flags, getRootDirectory());
+    return FileTable::get().add(openFile);
+  }
 
-  __wasi_size_t num = 0;
-  for (size_t i = 0; i < iovs_len; i++) {
-    const uint8_t* buf = iovs[i].buf;
-    __wasi_size_t len = iovs[i].buf_len;
+  long err;
+  std::shared_ptr<Directory> parentDir = getParent(pathParts, err);
 
-    lockedFile.read(buf, len);
-    num += len;
+  // Parent node doesn't exist.
+  if (!parentDir) {
+    return err;
   }
-  *nread = num;
-  return __WASI_ERRNO_INVAL;
+
+  auto lockedParentDir = parentDir->locked();
+
+  std::shared_ptr<File> curr =
+    lockedParentDir.getEntry(pathParts[pathParts.size() - 1]);
+
+  // The requested node was not found.
+  if (!curr) {
+    // If curr is the last element and the create flag is specified
+    if (flags & O_CREAT) {
+      // Create an empty in-memory file.
+      auto created = std::make_shared<MemoryFile>(mode);
+
+      lockedParentDir.setEntry(pathParts[pathParts.size() - 1], created);
+      auto openFile = std::make_shared<OpenFileState>(0, flags, created);
+
+      return FileTable::get().add(openFile);
+    } else {
+      return -(ENOENT);
+    }
+  }
+
+  // Fail if O_DIRECTORY is specified and pathname is not a directory
+  if (flags & O_DIRECTORY && !curr->is<Directory>()) {
+    return -(ENOTDIR);
+  }
+
+  // If the file exists and O_EXCL and O_CREAT are true
+  if (flags & O_EXCL && flags & O_CREAT) {
+    return -(EEXIST);
+  }
+
+  auto openFile = std::make_shared<OpenFileState>(0, flags, curr);
+
+  return FileTable::get().add(openFile);
 }
 
-__wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
-  int accessMode = (flags & O_ACCMODE);
-  bool canWrite = false;
+long __syscall_mkdir(long path, long mode) {
+  std::vector<std::string> pathParts = splitPath((char*)path);
```

Comment:
```suggestion
  auto pathParts = splitPath((char*)path);
```
I think these verbose types don't help readability, personally, in cases like this where it's clear we are returning a split path's parts.